File: src/output.rs
use std::borrow::Cow;
use std::io::{self, IsTerminal, Write};

use crate::ansi::strip_sgr;
use crate::config::{cfg, color_enabled, Pager};

pub fn write_output(buf: &[u8], _line_count: usize) -> io::Result<()> {
    let filtered: Cow<[u8]> = if color_enabled() {
        Cow::Borrowed(buf)
    } else {
        Cow::Owned(strip_sgr(buf))
    };
    let mut sink = OutputSink::open();
    sink.write_all(filtered.as_ref())?;
    sink.close()
}

// Output destination: either stdout or a spawned pager child. Shared by the
// buffered path (`write_output`) and the streaming path (`stream.rs`).
pub enum OutputSink {
    Stdout(io::Stdout),
    Child {
        child: std::process::Child,
        closed: bool,
    },
}

impl OutputSink {
    // Spawn a pager when configured + a usable PAGER is set, else write to
    // stdout. Auto only pages on a TTY; Always pages whenever PAGER is set;
    // Never never pages. A failed spawn falls back to stdout.
    pub fn open() -> Self {
        let is_tty = io::stdout().is_terminal();
        let pager_var = std::env::var("PAGER").ok().filter(|s| !s.trim().is_empty());
        let want_page = match cfg().pager {
            Pager::Never => false,
            Pager::Always => pager_var.is_some(),
            Pager::Auto => is_tty && pager_var.is_some(),
        };
        if want_page {
            if let Some(s) = pager_var {
                if let Some(sink) = spawn_pager(&s) {
                    return sink;
                }
            }
        }
        OutputSink::Stdout(io::stdout())
    }

    pub fn mark_closed(&mut self) {
        match self {
            OutputSink::Stdout(_) => {}
            OutputSink::Child { closed, .. } => *closed = true,
        }
    }

    pub fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
        match self {
            OutputSink::Stdout(stdout) => {
                let mut out = stdout.lock();
                out.write_all(buf)?;
                out.flush()
            }
            OutputSink::Child { child, closed } => {
                if *closed {
                    return Ok(());
                }
                let stdin = child.stdin.as_mut().expect("stdin piped");
                stdin.write_all(buf)?;
                stdin.flush()
            }
        }
    }

    pub fn close(mut self) -> io::Result<()> {
        match &mut self {
            OutputSink::Stdout(stdout) => {
                let mut out = stdout.lock();
                out.flush()?;
            }
            OutputSink::Child { child, closed: _ } => {
                drop(child.stdin.take());
                let _ = child.wait();
            }
        }
        Ok(())
    }
}

// Use std::process::Command (which calls posix_spawn on macOS) rather than
// a manual fork+exec. Calling fork() after the Rust runtime initializes the
// macOS frameworks/libdispatch can race the dispatch workqueue and get the
// process killed (SIGKILL) on Apple Silicon.
fn spawn_pager(cmd_line: &str) -> Option<OutputSink> {
    use std::process::{Command, Stdio};

    let mut parts = cmd_line.split_whitespace().map(|s| s.to_string());
    let cmd = parts.next()?;
    let args: Vec<String> = parts.collect();
    let _ = io::stdout().flush();
    let _ = io::stderr().flush();
    let child = Command::new(&cmd)
        .args(&args)
        .stdin(Stdio::piped())
        .spawn()
        .ok()?;
    Some(OutputSink::Child {
        child,
        closed: false,
    })
}

File: internal/output/output.go
// Package output owns the destination bijjou writes to: stdout or a spawned
// pager child.
package output

import (
	"io"
	"os"
	"os/exec"
	"strings"

	"github.com/charmbracelet/x/term"

	"tangled.org/jjt.io/bijjou/internal/ansi"
	"tangled.org/jjt.io/bijjou/internal/config"
)

// WriteOutput sends one finished buffer to the sink, stripping colour first
// when colour is off.
func WriteOutput(buf []byte) error {
	filtered := buf
	if !config.ColorEnabled() {
		filtered = ansi.StripSGR(buf)
	}
	sink := Open()
	if err := sink.Write(filtered); err != nil {
		return err
	}
	return sink.Close()
}

// Sink is an output destination: either stdout or a spawned pager child. The
// buffered path (WriteOutput) and the streaming path share it.
type Sink struct {
	// cmd is nil for the stdout sink.
	cmd    *exec.Cmd
	stdin  io.WriteCloser
	closed bool
}

// Open spawns a pager when bijjou is configured for one and PAGER holds a
// usable value. If not, it writes to stdout. The auto mode pages only on a
// TTY. The always mode pages when PAGER is set. The never mode does not page.
// A failed spawn falls back to stdout.
func Open() *Sink {
	isTTY := term.IsTerminal(os.Stdout.Fd())
	pagerVar := os.Getenv("PAGER")
	hasPager := strings.TrimSpace(pagerVar) != ""
	wantPage := false
	switch config.Get().Pager {
	case config.ModeNever:
		wantPage = false
	case config.ModeAlways:
		wantPage = hasPager
	case config.ModeAuto:
		wantPage = isTTY && hasPager
	}
	if wantPage {
		if sink := spawnPager(pagerVar); sink != nil {
			return sink
		}
	}
	return &Sink{}
}

// MarkClosed records that the child stopped reading. Later writes are drops.
func (s *Sink) MarkClosed() {
	if s.cmd != nil {
		s.closed = true
	}
}

// Write sends buf to the sink.
func (s *Sink) Write(p []byte) error {
	if s.cmd == nil {
		_, err := os.Stdout.Write(p)
		return err
	}
	if s.closed {
		return nil
	}
	_, err := s.stdin.Write(p)
	return err
}

// Close flushes stdout, or closes the child stdin and reaps the pager.
func (s *Sink) Close() error {
	if s.cmd == nil {
		return nil
	}
	if s.stdin != nil {
		_ = s.stdin.Close()
		s.stdin = nil
	}
	_ = s.cmd.Wait()
	return nil
}

// spawnPager starts the pager named by cmdLine with a piped stdin. It returns
// nil when the command line is empty or the spawn fails.
func spawnPager(cmdLine string) *Sink {
	parts := strings.Fields(cmdLine)
	if len(parts) == 0 {
		return nil
	}
	cmd := exec.Command(parts[0], parts[1:]...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil
	}
	if err := cmd.Start(); err != nil {
		_ = stdin.Close()
		return nil
	}
	return &Sink{cmd: cmd, stdin: stdin}
}
